# Credit Risk Classification

Credit risk poses a classification problem that’s inherently imbalanced. This is because healthy loans easily outnumber risky loans. In this Challenge, you’ll use various techniques to train and evaluate models with imbalanced classes. You’ll use a dataset of historical lending activity from a peer-to-peer lending services company to build a model that can identify the creditworthiness of borrowers.

## Instructions:

This challenge consists of the following subsections:

* Split the Data into Training and Testing Sets

* Create a Logistic Regression Model with the Original Data

* Predict a Logistic Regression Model with Resampled Training Data 

### Split the Data into Training and Testing Sets

Open the starter code notebook and then use it to complete the following steps.

1. Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

2. Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

    > **Note** A value of `0` in the “loan_status” column means that the loan is healthy. A value of `1` means that the loan has a high risk of defaulting.  

3. Check the balance of the labels variable (`y`) by using the `value_counts` function.

4. Split the data into training and testing datasets by using `train_test_split`.

### Create a Logistic Regression Model with the Original Data

Employ your knowledge of logistic regression to complete the following steps:

1. Fit a logistic regression model by using the training data (`X_train` and `y_train`).

2. Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.

4. Answer the following question: How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Predict a Logistic Regression Model with Resampled Training Data

Did you notice the small number of high-risk loan labels? Perhaps, a model that uses resampled data will perform better. You’ll thus resample the training data and then reevaluate the model. Specifically, you’ll use `RandomOverSampler`.

To do so, complete the following steps:

1. Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

2. Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.
    
4. Answer the following question: How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Write a Credit Risk Analysis Report

For this section, you’ll write a brief report that includes a summary and an analysis of the performance of both machine learning models that you used in this challenge. You should write this report as the `README.md` file included in your GitHub repository.

Structure your report by using the report template that `Starter_Code.zip` includes, and make sure that it contains the following:

1. An overview of the analysis: Explain the purpose of this analysis.


2. The results: Using bulleted lists, describe the balanced accuracy scores and the precision and recall scores of both machine learning models.

3. A summary: Summarize the results from the machine learning models. Compare the two versions of the dataset predictions. Include your recommendation for the model to use, if any, on the original vs. the resampled data. If you don’t recommend either model, justify your reasoning.

In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

import warnings
warnings.filterwarnings('ignore')

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
# YOUR CODE HERE!

# Review the DataFrame
# YOUR CODE HERE!

In [3]:
file_path = Path("/Users/theo/Desktop/Local_Workspace/TRG_Challenge_12/Resources/lending_data.csv")
df = pd.read_csv(file_path)

In [4]:
df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

In [5]:
df.tail()

loan_size  interest_rate  borrower_income  debt_to_income  \
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [6]:
# Separate the data into labels and features

# Separate the y variable, the labels
# YOUR CODE HERE!]

# Separate the X variable, the features
# YOUR CODE HERE!

In [12]:
# Split target column from dataset
target = df['loan_status']
features = df[['loan_size','interest_rate','borrower_income','debt_to_income','num_of_accounts','derogatory_marks','total_debt']]


In [13]:
# Review the y variable Series
# YOUR CODE HERE!

target[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [14]:
# Review the X variable DataFrame
# YOUR CODE HERE!

features = pd.get_dummies(X)

features.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [10]:
# Check the balance of our target values
# YOUR CODE HERE!

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [15]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
# YOUR CODE HERE!

training_features, testing_features, training_targets, testing_targets = train_test_split(features, target)


In [16]:
training_features

loan_size  interest_rate  borrower_income  debt_to_income  \
64077     7900.0          6.485            41600        0.278846   
36543     9500.0          7.150            47900        0.373695   
21560    11300.0          7.938            55300        0.457505   
74207    11200.0          7.869            54600        0.450549   
70878     9900.0          7.327            49500        0.393939   
...          ...            ...              ...             ...   
26734     9100.0          7.005            46500        0.354839   
6116      9000.0          6.947            46000        0.347826   
42231    10000.0          7.368            49900        0.398798   
26628     8800.0          6.865            45200        0.336283   
28715     9700.0          7.268            49000        0.387755   

       num_of_accounts  derogatory_marks  total_debt  
64077                2                 0       11600  
36543                4                 0       17900  
21560                5                 1       25300  
74207                5                 1       24600  
70878                4                 0       19500  
...                ...               ...         ...  
26734                3                 0       16500  
6116                 3                 0       16000  
42231                4                 0       19900  
26628                3                 0       15200  
28715                4                 0       19000  

[58152 rows x 7 columns]

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [28]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!

# Fit the model using training data
# YOUR CODE HERE!

In [32]:
logistic_regression_model = LogisticRegression()

In [33]:
logistic_regression_model.fit(training_features, training_targets)
LogisticRegression()


LogisticRegression()

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [37]:
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

In [ ]:
# Separate the data into y and X
#y = df["y"]
# The "X" column should be all the features
#X = df.drop(columns="y")


In [ ]:
y = df['loan_status']
X = df[['loan_size','interest_rate','borrower_income','debt_to_income','num_of_accounts','derogatory_marks','total_debt']]


In [38]:
# Split into testing and training sets
X_train, X_test, y_train, y_test = train_test_split(X, y)

# Train the logistic regression classifier
logistic_regression_model = LogisticRegression()

# Then fit it
logistic_regression_model.fit(X_train, y_train)

LogisticRegression()

In [39]:
#Generate training predictions
training_predictions = logistic_regression_model.predict(X_train)


In [40]:
# Make a prediction using the testing data
# YOUR CODE HERE!

testing_predictions = logistic_regression_model.predict(X_test)


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [41]:
# Print the balanced_accuracy score of the model
# YOUR CODE HERE!

# Import the accuracy_score function
from sklearn.metrics import accuracy_score

# Calculate the model's accuracy on the test dataset
accuracy_score(testing_targets, testing_predictions)

0.9353590590177466

In [42]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!

#Import the module
from sklearn.metrics import confusion_matrix

#Print the confusion matrix
training_matrix = confusion_matrix(y_train, training_predictions)
print(training_matrix)

[[55992   289]
 [  174  1697]]


In [43]:
# Print the classification report for the model
# YOUR CODE HERE!

# Import the module
from sklearn.metrics import classification_report
# Save a classification report
training_report = classification_report(y_train, training_predictions)
# Then view its output
print(training_report)

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     56281
           1       0.85      0.91      0.88      1871

    accuracy                           0.99     58152
   macro avg       0.93      0.95      0.94     58152
weighted avg       0.99      0.99      0.99     58152



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** WRITE YOUR ANSWER HERE!

** The logistic regression model has a high accuracy rate .99 with a 1.00 on healthy loans and .85 on high risk loans.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [44]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!

# Fit the original training data to the random_oversampler model
# YOUR CODE HERE!

In [ ]:
# Count the distinct values of the resampled labels data
# YOUR CODE HERE!

In [45]:
random_oversampler = RandomOverSampler(random_state=1)
X_resampled, y_resampled = random_oversampler.fit_resample(X_train, y_train)

In [47]:
y_resampled.value_counts()

0    56281
1    56281
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [ ]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!


# Fit the model using the resampled training data
# YOUR CODE HERE!

# Make a prediction using the testing data
# YOUR CODE HERE!

In [48]:
# Declare a logistic regression model
model = LogisticRegression(random_state=1)

# Fit the model
model.fit(X_resampled, y_resampled)

LogisticRegression(random_state=1)

In [49]:
# Predict the labels for the testing features
y_pred = model.predict(X_test)

# Print the confusion matrix
print(confusion_matrix(y_test, y_pred))

[[18654   101]
 [    5   624]]


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [50]:
# Print the balanced_accuracy score of the model 
# YOUR CODE HERE!

# Import the module
from sklearn.metrics import balanced_accuracy_score

# Print the accuracy score
balanced_accuracy_score(y_test, y_pred)

0.9933328218993218

In [54]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!

# Predict the labels for the testing features
y_pred = model.predict(X_test)

# Print the confusion matrix
print(confusion_matrix(y_test, y_pred))

[[18654   101]
 [    5   624]]


In [52]:
# Print the classification report for the model
# YOUR CODE HERE!

# Import the module
from imblearn.metrics import classification_report_imbalanced

# Print the classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

          0       1.00      0.99      0.99      1.00      0.99      0.99     18755
          1       0.86      0.99      0.99      0.92      0.99      0.99       629

avg / total       1.00      0.99      0.99      0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** YOUR ANSWER HERE!

** The logistic regression model fits the oversampled data with 1.00 on healthy loans and .86 on high risk loans with an accuracy of .99. 